atp/client.go

```diff
@@ -235,14 +244,41 @@ func (c *client) Close() error {
 			clientDoneMessage{},
 		})
 		if err != nil {
-			return fmt.Errorf("client with steps '%s' failed to write client done message with error: %w",
-				c.getRunningStepIDs(), err)
+			// add a timeout to the wait to prevent it from causing a deadlock.
+			// 5 seconds is arbitrary, but gives it enough time to exit.
+			waitedGracefully := waitWithTimeout(time.Second*5, &c.wg)
+			if waitedGracefully {
+				return fmt.Errorf("client with step '%s' failed to write client done message with error: %w",
+					c.getRunningStepIDs(), err)
+			} else {
+				panic(fmt.Errorf("potential deadlock after client with step '%s' failed to write client done message with error: %w",
+					c.getRunningStepIDs(), err))
+			}
 		}
 	}
 	c.wg.Wait()
 	return nil
 }
```

Comment:
Is this proposing to merge the to wait calls? I think the variable, while unnecessary, adds context.
I haven't found any potential cases that cause the wait to fail in the non-error case. So there should be no timeout on that.

Reply:
If there isn't an error, it's better to not have a timeout. The timeout is just to prevent it from hanging due to an error. Errors can cause problems that cause it to hang. I don't recall it ever hanging when there wasn't an error.

Reply:
In [a conversation below](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/92#discussion_r1661469274), you are advocating avoiding an unbounded wait, and here you're doing the opposite.  😀

We don't want to have the wait time out, but we also don't want to wait forever; and, that's the case regardless of whether an error occurred or not.  If you can put a reasonableness bound on the wait in the non-error case (like you've done here for the error case), then there is no reason not to use a timeout (now that we have a nice function for it).  And, if it's reasonable to use a timeout in both cases, then you might as well use common code (because is cleaner, simpler, easier to test, etc.).

Reply:
In the linked case, I can guarantee that, as designed, a second wait call would hang it until the process is terminated. In this case, I don't want to make assumptions that could risk exiting early. I've never had it hang on the happy path as designed, and if that changes, we can always change this in the future, either while debugging or permanently.